File: Python/PathFinding/Pathfinder.py
import pygame
from MazeObjects.Colors import Color
from PathFinding.PathfindingTile import Tile


class Pathfinder:
    def __init__(self, grid, goal):
        self.x_coordinate, self.y_coordinate = 0, 0
        self.goal_x_coordinate, self.goal_y_coordinate = goal
        self.open_list, self.closed_list = [], []
        self.maze = grid
        self.grid = [Tile((cell.x_coordinate, cell.y_coordinate)) for cell in grid]
        self.tile = self.grid[0]
        self.tile.score(-1, (9, 9))

    def is_passable(self, position, target, grid_width):
        """
        Checks if the pathfinder's chosen neighbor is not barricaded by walls
        :param position: The directional position of the chosen neighbor
        :param target: The index of the chosen neighbor in the grid array
        :param grid_width: The number of columns in the grid
        :return: None
        """
        curr = self.x_coordinate + self.y_coordinate * grid_width
        if position is 'bottom':
            if self.maze[curr].walls[position].show is True and \
                    self.maze[target].walls['top'].show is True:
                    return True
        elif position is 'top':
            if self.maze[curr].walls[position].show is True and \
                    self.maze[target].walls['bottom'].show is True:
                    return True
        elif position is 'right':
            if self.maze[curr].walls[position].show is True and \
                    self.maze[target].walls['left'].show is True:
                    return True
        else:
            if self.maze[curr].walls[position].show is True and \
                    self.maze[target].walls['right'].show is True:
                    return True

        return False

    def path_find(self, grid_width=10):
        """
        Makes the path finder find the best path to the target
        :return: None
        """
        neighbors = {'top': (self.x_coordinate, self.y_coordinate - 1),
                     'bottom': (self.x_coordinate, self.y_coordinate + 1),
                     'left': (self.x_coordinate - 1, self.y_coordinate),
                     'right': (self.x_coordinate + 1, self.y_coordinate)}

        for neighbor in neighbors:
            coordinates = neighbors[neighbor]
            neighbor_idx = coordinates[0] + coordinates[1] * grid_width
            if coordinates[0] < 0 or coordinates[1] < 0 \
                    or coordinates[0] > grid_width - 1 or coordinates[1] > grid_width - 1\
                    or self.is_passable(neighbor, neighbor_idx, grid_width) \
                    or self.grid[neighbor_idx] in self.closed_list:
                continue

            curr = self.grid[neighbor_idx]
            curr.score(self.tile.g_score, (self.goal_x_coordinate, self.goal_y_coordinate))
            self.open_list.append(curr)

        # sort the tiles in the open list by f score in ascending order
        # so we always get the tile with the smallest possible f score
        self.open_list.sort(key=lambda n: n.f_score)

        if self.open_list:
            chosen = self.open_list[0]
            parent = self.tile
            chosen.parent = parent
            self.tile = chosen
            self.x_coordinate, self.y_coordinate = chosen.x_coordinate, chosen.y_coordinate
            self.closed_list.append(chosen)
            self.open_list.clear()
        else:
            chosen = self.tile.parent
            self.tile = chosen
            self.x_coordinate, self.y_coordinate = chosen.x_coordinate, chosen.y_coordinate

File: JavaScript/PathFinderObjects/PathFinder.js
function PathFinder(x_coordinate, y_coordinate, grid){
  this.x_coordinate = x_coordinate;
  this.y_coordinate = y_coordinate;
  this.open_list = [];
  this.closed_list = [];
  this.maze = grid;
  this.grid = grid.map(cell => new PathFindingCell(cell.x_coordinate,
    cell.y_coordinate));
  this.curr_cell = this.grid[this.GetIndex(this.x_coordinate,
    this.y_coordinate, this.grid[this.grid.length - 1].x_coordinate)];
  this.goal = this.grid[this.grid.length - 1];
  this.curr_cell.ComputeScore(-1, this.goal);
}


PathFinder.prototype.GetIndex = function (x_coordinate, y_coordinate, grid_width) {
  if (x_coordinate < 0 || y_coordinate < 0 || x_coordinate > grid_width - 1
    || y_coordinate > grid_width - 1) {
      return - 1;
    }
  else {
    return x_coordinate + y_coordinate * grid_width;
  }
};

PathFinder.prototype.IsPassable = function (x_coordinate, y_coordinate, grid_width) {
  let cur = this.maze[this.x_coordinate + this.y_coordinate * grid_width].Walls;
  let neighbor = this.maze[x_coordinate + y_coordinate * grid_width].Walls;
  let n = this.maze[x_coordinate + y_coordinate * grid_width];
  // n.Highlight(255, 255, 255);
  // console.log(neighbor);
  if ((x_coordinate - this.x_coordinate) == 1) {
    if (!cur["right"].show && !neighbor["left"].show){
      return true;
    }
  }
  else if ((x_coordinate - this.x_coordinate) == -1){
    if (!cur["left"].show && !neighbor["right"].show){
      return true;
    }
  }
  else if ((this.y_coordinate - y_coordinate) == 1) {
    if (!cur["top"].show && !neighbor["bottom"].show){
      return true;
    }
  }
  else {
    if (!cur["bottom"].show && !neighbor["top"].show){
      return true;
    }
  }

  // return false;
};


PathFinder.prototype.PathFind = function (grid_width) {
  let indexes = [
    this.GetIndex(this.x_coordinate, this.y_coordinate + 1, grid_width),
    this.GetIndex(this.x_coordinate, this.y_coordinate - 1, grid_width),
    this.GetIndex(this.x_coordinate + 1, this.y_coordinate, grid_width),
    this.GetIndex(this.x_coordinate - 1, this.y_coordinate, grid_width)
  ];




  indexes = indexes.filter(index => index != -1 && index != 0);

  for (let i = 0; i < indexes.length; i++) {
    if (!this.closed_list.includes(this.grid[indexes[i]])
        && !this.open_list.includes(indexes[i])) {
      this.open_list.push(indexes[i]);
    }
  }

  this.open_list = this.open_list.filter(index => index != -1);
  // console.log(this.open_list);

  let neighbors = [];

  for (let i = 0; i < this.open_list.length; i++) {
    let n = this.grid[this.open_list[i]];
    if (!neighbors.includes(this.grid[this.open_list[i]])
        && this.IsPassable(n.x_coordinate, n.y_coordinate, grid_width)) {
          n.ComputeScore(this.curr_cell.G_score, this.goal);
          neighbors.push(n);
    }
  }

  // neighbors = neighbors.sort((a, b) => a.F_score - b.F_score);

  // console.log(neighbors);

  let chosen = undefined;

  if (neighbors.length > 0) {
    chosen = neighbors[0];
  }

  // if (this.open_list.length > 0) {
  //   for (var i = 0; i < this.open_list.length; i++) {
  //     let nei = this.maze[this.open_list[i]];
  //     this.grid[this.open_list[i]].ComputeScore(this.curr_cell.G_score, this.goal);
  //     if (this.grid[this.open_list[i]].F_score <= this.curr_cell.F_score
  //       && this.IsPassable(nei.x_coordinate, nei.y_coordinate, grid_width)) {
  //         // console.log(nei);
  //         chosen = this.grid[this.open_list[i]];
  //     }
  //   }
  // }

  // console.log(chosen);
  if (chosen != undefined) {
    this.x_coordinate = chosen.x_coordinate;
    this.y_coordinate = chosen.y_coordinate;
    this.curr_cell = chosen;
    this.closed_list.push(chosen);
    this.open_list = [];
  }
  else {

  }


};
